### DISCUSSION TITLE
Markdown and Razor

### DISCUSSION THREAD

**Discussion by ror3d** [statiqdev/Discussions#109] [Q&A] [answered]
I've read in the documentation that `.md` files are processed by Markdown and then Razor. I expected that this would mean I can use `@` razor commands, like conditionals, or defining a @section. But it doesn't seem to work, I think Markdown is converting the @ to `&#64;`, and also any `{ ... }` into tag attributes.

Is there some way this can be done?

Thanks

**Comment by daveaglick** [comment, marked as answer]
> I think Markdown is converting the @ to &#64;

That's correct. The raw `RenderMarkdown` module has a toggle for this behavior, but in Statiq Web it's configured to HTML escape a single `@` symbol since we know Razor will be run on the rendered Markdown afterwards. This is a little tricky because without that escaping we were getting all kinds of problem reports of folks not understanding why things like email addresses in their Markdown were triggering errors when processed by Razor.

The good news is that it's easy to un-escape the `@` HTML escaping using `\@` in your Markdown. I.e. `\@DateTime.Now.ToString()`.

Let me know if you have any trouble with that.

**Reply by ror3d** [reply]
I might be getting something wrong, but the `\@DateTime.Now.ToString()` does not work for me. I'm just adding it to a markdown file, and the output in the HTML is still `&#64;DateTime.Now.ToString()`.

What *does* work is escaping the `\{`, so I think that if the `@` worked, something like 

    \@section Nav \{
        some content
    }

maybe would do the trick.

**Reply by daveaglick** [reply]
Finally had a chance to play around with this a bit, and was a little surprised that I could replicate your results. After a little trial-and-error I realized what was going on.

When `\@something` is written in Markdown, the first `\` is "consumed" by the Markdown engine to indicate that the `@` should not be escaped. That's all basically [according to the CommonMark specification](https://spec.commonmark.org/0.30/#backslash-escapes). So before Statiq even gets to it, the Markdown renderer sees `\@`, treats the `\` as an escape for the `@`, and outputs a single `@`. Then because Statiq looks for `@` and HTML encodes them by default, it sees the bare `@` (since the Markdown renderer are our `\`) and encodes it.

You can get around this in two ways today:
- Write two slashes so that even when the Markdown renderer consumes the first one, the second one is then used by Statiq to escape the HTML encoding: `\\@DateTime.Now.ToString()`.
- Alternatively, you can put the Razor code inside an HTML element which [doesn't do the normal Markdown escaping behavior](https://spec.commonmark.org/0.30/#raw-html) and therefore won't consume the slash: `<div>\@DateTime.Now.ToString()</div>`.

The problem here is that it's not as advertised and it's confusing. You were right on to see a problem here. I'm going to see if I can hook the Markdown engine so that `@` symbols are HTML encoded _inside_ the Markdown processing instead of in a second pass. That should help standardize everything and work as expected. I'll update here depending on what I end up being able to do.

**Reply by ror3d** [reply]
Oh awesome! Thanks so much for looking into it. It's good there's at least workaround for this.

Maybe a simple hack to get it to have the expected behaviour would be to have a step in the markdown previous to calling it that converts `\@` to `\\@`. I don't know if this could have side-effects.

**Reply by daveaglick** [reply]
This is now fully resolved. I initially started with a string search-and-replace like you suggested (that's actually how it worked in the first place), but the way Markdown handles escape characters is tricky, especially since it's different depending on whether you're inside a block like HTML or a code fence or in "normal" Markdown. Finally ended up with a solution that works by hooking the Markdig engine with a custom parser that intercepts escapes when, and only when, they're followed by an `@`.

Expect this to work better in the next release, but be warned that you _may_ need to adjust any `\\@` you added as a work-around (though in an interesting quirk, that might still work but not for the reasons you might expect - the `\\` will be seen by the Markdown processor as "escape the slash that would normally be interpreted as escaping the `@` and output it directly", which the special post-process phase in Statiq that encodes the `@` symbols will see as `\@` and therefore still encode it).

**Reply by ror3d** [reply]
Cool! thank you so much!

**Reply by ror3d** [reply]
Hi, sorry, having more trouble with this stuff (although maybe there's some alternative to what I'm trying that I haven't found).

I was trying to use a `@Document.GetString("name")` inside a markdown, and the only way I got it to work was by using it inside a `<div>@Document.GetString("name")</div>`. Basically, if I try just the `@` command, it fails because it encodes the `"` as `&quot" and then breaks because of that.

Also, a note on the "hack" of using it inside an html tag, it seems to work with `<div>`, but it does not work with `<span>` for example.

**Reply by daveaglick** [reply]
This is all sort of circling around the problem that Markdown->HTML->Razor->HTML is pretty tricky to get stuff to flow through correctly. Have you taken a look at [shortcodes](https://www.statiq.dev/guide/content-and-data/shortcodes)? That feature is designed to provide something like a universal templating mechanism that works directly in Markdown.

For example, to output the value of the "name" metadata, you write this inside the Markdown file:

```
<?!= Name /?>
```

The `<?!=` syntax is something I need to document better 😬. That might be more in line with what you're looking to do.

**Reply by ror3d** [reply]
Ah right! yes, that's exactly what I wanted! I looked into the shortcodes page looking to see if there was something like this but I didn't find it. Thank you!

**Reply by daveaglick** [reply]
> I looked into the shortcodes page looking to see if there was something like this but I didn't find it

Because it's not there! That's on me - still trying to get everything that's in here documented after all these years working on it. I'm easily distracted.

![distracted](https://media.giphy.com/media/Mt0IKnQaKdSTu/giphy.gif)

**Reply by ror3d** [reply]
That's fine, good to know, and it stays here for reference while it's not added there I guess :) Thanks a lot for your work, even though there's some rough edges, this system is one of the best out there for what it does
